### ISSUE TITLE
Generator functions for ES6 [FCC BETA]

### ISSUE THREAD

**Issue by Flarp** [freeCodeCamp/freeCodeCamp#14234]
Generator functions are a big feature in ES6, and the building blocks for the new async/await feature in ES201(6/7). It should certainly be covered in some way, like an introduction, and maybe a tutorial on how to make a version of async/await using generators. 

Something like

* Generators
* Some generator practice
* Async/await using generators

**Comment by HKuz** [comment]
@Flarp - thanks for noting these topics, especially in light of the future of JS. I remember that generators, at least, were brought up as potential challenges in the ES6 section during the curriculum expansion efforts. (If you're curious, you can check out that [old issue here](https://github.com/freeCodeCamp/CurriculumExpansion/issues/34)). Unfortunately, some of those challenges were never written.

If you (or any other contributor) is interested in creating a set of fCC-style interactive challenges to cover these topics in the ES6 section, that would be awesome! Check out the [challenge style guide](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/seed/challenge-style-guide.md) and feel free to ask any questions or get guidance in the [Contributor's chat room](https://gitter.im/FreeCodeCamp/Contributors).

**Head's up**: there's a known issue in ES6 challenge tests. The camper's code is transpiled into ES5 syntax BEFORE the tests are run on the Beta site, so any tests using regex to check for specific ES6 keywords, etc. won't work.

**Comment by Flarp** [comment]
I won't be able to work on this until summer, but I am open to working on it

**Comment by alan-luo** [comment]
I would like to tackle this if @Flarp is busy.

**Comment by Flarp** [comment]
I am a little busy, but I would be glad to work with you.

**Comment by Flarp** [comment]
It seems @alan-luo is busy, and @QuincyLarson wants this done for October. I'll do my best to start this over the weekend.

**Comment by alan-luo** [comment]
@Flarp Yes, I underestimated how much work I'd be getting this year! Will try to help if at all possible.

**Comment by QuincyLarson** [comment]
@Flarp @alan-luo Ah, to be clear, I added the Hacktoberfest label so people would see it as an issue they could potentially contribute to. It does not have a specific deadline in mind.

**Comment by Flarp** [comment]
I will probably not be able to work on this unless I can get help from someone.

**Comment by sblack4** [comment]
@Flarp do you have an idea for a curriculum? I feel like the generator lesson could be fun

**Comment by Flarp** [comment]
@sblack4 How to make one, how to use one, and real world applications. That's my idea, anyways.

**Comment by sblack4** [comment]
@Flarp I'm down, that sounds like fun

**Comment by sblack4** [comment]
@Flarp , @HKuz Do you know if we [still can't use ES6](https://github.com/freeCodeCamp/freeCodeCamp/issues/14234#issuecomment-292021710) ?

**Comment by Flarp** [comment]
@sblack4 Awesome. Before I go any further, I should run a refresher with iterators. I haven't touched JS in months.

**Comment by sblack4** [comment]
@Flarp javascript is like riding a bike

**Comment by Flarp** [comment]
@sblack4 is there a punchline

**Comment by sblack4** [comment]
@Flarp hold on... when I think of one I'll let you know

**Comment by Flarp** [comment]
@sblack4 email me when you get one (and so we can have communication without spamming the issue)

**Comment by HKuz** [comment]
Hi @Flarp and @sblack4 - unfortunately, I'm not sure about the ES6 issue, I've been out of the loop on the platform updates. Maybe try asking in the chat room, and check if anyone else knows. Cheers for your renewed interest in creating these challenges!

**Comment by QuincyLarson** [comment]
@Flarp We just launched the new freeCodeCamp curriculum. Would you still be interested in helping design a few challenges that introduce these concepts?

**Comment by Flarp** [comment]
@QuincyLarson I have finals in a few days, and I would love to do it, but it would be a little while

**Comment by QuincyLarson** [comment]
@Flarp Good luck with your finals! 

Well, we've waited this long. Can you can commit to fixing this after you finish your finals?

**Comment by Flarp** [comment]
@QuincyLarson I will do my absolute best

**Comment by Flarp** [comment]
@QuincyLarson I will be working on it from now until June 14th, and after June 22nd.

**Comment by Flarp** [comment]
Blocked (?) by #17510

**Comment by Flarp** [comment]
As stated above, I will be gone for a week, but more than willing to get started once I return.

**Comment by QuincyLarson** [comment]
@Flarp OK - have a nice week off. Hopefully we'll have you un-blocked by the time you get back :)

**Comment by Flarp** [comment]
Generators and iterators are similar enough to be grouped into the same project. Map and Set are also important, and use these tactics to be formulated. My game plan for this is to re-introduce the concept of iteration, and then introduce for..of, Maps, Sets, and how to make an iterator yourself. The second part should be generators, going over the concept of scopes and closures again, and then going into some simple use cases, how generators are also iterators, and then some real world examples of a generator.

**Comment by Flarp** [comment]
I've started my work [here](https://github.com/Flarp/curriculum), feedback is appreciated.

**Comment by Flarp** [comment]
https://github.com/freeCodeCamp/curriculum/issues/68 Restarting `learn` takes about 7 minutes at a time, and I can't find another way to test my changes in `curriculum` as of now. Is there some way I can do it without having to wait such a long time?

**Comment by Flarp** [comment]
Here's another block - https://github.com/freeCodeCamp/curriculum/issues/42

Due to not being able to work on this in continuous chunks, and no resolution to any of these as of yet, I can't really work on this as of now. I am not familiar enough with the codebase to be able to fix the first issue, but I might be able to tackle the second one.

**Comment by QuincyLarson** [comment]
@Flarp Thanks for your extreme patience with my slow response. I'm glad we've managed to fix some of the issues that were impeding your progress. Looking at your latest commit it's hard for me to untangle exactly which challenges you've created. When we deploy and QA these we may want to just create a single PR that adds the new challenges to their corresponding JSON files.

How many new challenges have you created, and how soon can we start working through them and giving feedback? :)

**Comment by Flarp** [comment]
I have been creating challenges locally, just not pushing them. I have three out of the seven challenges complete, there was a lot of setbacks in my real-life, but I think I can get almost everything done this week.

**Comment by QuincyLarson** [comment]
@Flarp Awesome! As soon as you're ready for us to give feedback and QA, let us know. Eager to get these new challenges live :)

**Comment by Flarp** [comment]
I believe it should be ready for peer reviewing, the commits are not properly squashed but I can do that.

The indentation is strange due to some strange modifications in packing that I didn't make, but I hope it's satisfactory. I believe everything works, but this is a draft, not a final proposal. 

[Here you go](https://github.com/Flarp/curriculum)

**Comment by Flarp** [comment]
Any updates?

**Comment by QuincyLarson** [comment]
@Flarp Thanks for your patience.

@beaucarnes could you get this running locally and take a look at the new ES6-related challenges and make sure you can get follow the directions and get them to pass? Also please see if you can make the instructions and test descriptions as clear and concise as possible.

**Comment by jzeltn** [comment]
I can't sleep so I'll give it a try.

**Comment by Flarp** [comment]
@jzeltn Check the repo I linked, I think everything should already work.

**Comment by Flarp** [comment]
Any updates? I'd really like to hear back from someone.

**Comment by QuincyLarson** [comment]
@Flarp Thanks for your extreme patience. OK - so just to be clear, you are no longer blocked and your proposed changes are located here? https://github.com/Flarp/curriculum

Please confirm the exact files you'd like us to look at - preferably by opening a pull request that mentions this issue number.

**Comment by Flarp** [comment]
It's been a while since I've looked at the code. I need to re-examine and retest everything; there might be some test cases that are still broken. 

The only thing that blocked me was being unable to use unpacking. I eventually just gave up and edited the JSON raw, which is why I'm not 100% sure if the test cases work.

**Comment by QuincyLarson** [comment]
@Flarp OK - thanks for the update, and for taking time to re-examine the code. Please keep us posted, and let us know if we can do anything to help.

**Comment by Flarp** [comment]
It appears that the way you contribute to the fcc curriculum has changed; how can I translate my current codebase to the new standard?

Also, unfortunately, the laptop I have been using for most of my development has died, so it will take some time to find a suitable replacement and finally get the tests running.

**Comment by QuincyLarson** [comment]
@Flarp You should be able to pull down the latest master branch. This said, it may be easiest to just delete your local installation of freeCodeCamp and clone it again.

**Comment by Flarp** [comment]
Sorry for taking so long to respond; the next semester started and I was sidetracked with work. 

I'm going to work for a few hours today to make sure everything is in order.

**Comment by Flarp** [comment]
Unfortunately my 2010 netbook cannot handle having a web browser and Gatsby running at the same time, so I'm going to have to rewrite a lot of this blindly. If someone here could run the tests I make that would be great, because I can't.

**Comment by QuincyLarson** [comment]
@Flarp No worries - if you can make your changes and push your edits, one of us can QA them locally. I know it's tough to fly blind like that, so my hats off to you. Best of luck with it and thanks for keeping us posted!

**Comment by thecodingaviator** [comment]
@Flarp We'd be glad to test if you could go ahead and make PRs. Add a `Refer to #14234` in every PRs description please

**Comment by Flarp** [comment]
@thecodingaviator I've been really busy with AP classes, but I'll try to have something out ASAP

**Comment by Flarp** [comment]
Unfortunately I can't work on this anymore due to quite a bit of stuff going on in my life, but hopefully what I have is a good enough outline for continuing this (most of the articles are written, and while I'm not sure the tests work, the ideas are there). Sorry for going so long without a reply, life comes like a truck.

**Comment by thecodingaviator** [comment]
@Flarp I think we understand. Where can we find the work you've done and were referring to?

**Comment by Flarp** [comment]
https://github.com/Flarp/curriculum and https://github.com/Flarp/freeCodeCamp
